Implement collision preview

white-axe opened the pull request:
This pull request resolves #22 by implementing a new shader that renders on top of the map and tilepicker to visualize the collisions on the map. It can be toggled using a new entry at the bottom of the layer dropdown. It only takes into account the map layers that are enabled, disabled map layers do not contribute to the collision preview.

The shader is WebGL-compatible when combined with the changes from #52.

melody-rs commented on src/graphics/primitives/collision/shader.rs:
```diff
@@ -0,0 +1,155 @@
+                .device
+                .create_pipeline_layout(&wgpu::PipelineLayoutDescriptor {
+                    label: Some("Tilemap Collision Render Pipeline Layout (push constants)"),
+                    bind_group_layouts: &[&DUMMY_LAYOUT],
```
Why is there a dummy layout?

white-axe replied:
I was using that for testing purposes because the viewport is hardcoded to bind to group 1 and then didn't remove it. I've just changed the viewport to allow specifying a group index.

white-axe committed:
Add collision shader

This adds a new shader for collision that renders over the fog. It
doesn't do anything yet because the collision is not being calculated
anywhere yet.

(cherry picked from commit eeecc6cda340ed83a032182c67396969b021ecf6)

white-axe committed:
Fix loop dimensions in collision shader `calculate_vertices`

(cherry picked from commit 93bc2c93518d6d3d2221ad50877a6ba0a724c434)

white-axe committed:
Render fog and collision on top of event graphics

I also made the event border rectangles render on top of the fog and
collision because otherwise they may be hard to see through those
layers.

(cherry picked from commit ade0fa2a1ae36d8c42c56f985b46798b498f8331)

white-axe committed:
Start implementing collision calculation

(cherry picked from commit 7938c2226048132471a74667e37f17bc5b621365)

white-axe committed:
Replace image cache with dummy bind group in collision shader

(cherry picked from commit cbbf696d102af03d83edcd9a50ac0ce9645181c9)

white-axe committed:
Update collision preview when map changes

(cherry picked from commit 2ad9e87a46c136821391e7d8f707fb98894b3bb9)

white-axe committed:
Fix fog and collision not rendering when events are disabled

(cherry picked from commit b47e70b8eff64e0905d3aa77217eacbee4596b14)

white-axe committed:
Add collision preview to tilepicker as well

(cherry picked from commit c5dd6d01ea2e99951c3e6f9c7630bb05f0809331)

white-axe committed:
Collision preview now ignores disabled map layers

(cherry picked from commit 28e17bfac9b4d755b843afc1a0b62e9402dc659c)

white-axe committed:
Fix labels in layer picker not aligning with checkboxes

(cherry picked from commit e7770573d97ba17c466639e29d5439c07781a4d3)

white-axe committed:
Bind viewport to group 0 instead of using dummy layout

(cherry picked from commit 74874f2f552d8b0b4137d174b805884ed521cc92)

melody-rs commented on crates/components/src/tilepicker.rs:
```diff
@@ -145,8 +157,24 @@ impl Tilepicker {
+            luminol_data::Table2::new(tilepicker_data.xsize(), tilepicker_data.ysize());
+        for (y, x) in (0..tilepicker_data.ysize()).cartesian_product(0..tilepicker_data.xsize()) {
+            passages[(x, y)] =
+                tileset.passages[tilepicker_data[(x, y, 0)].try_into().unwrap_or_default()];
```
I'm getting a panic here:
```
The application panicked (crashed).
Message:  index out of bounds: the len is 1168 but the index is 1168
Location: crates\components\src\tilepicker.rs:164
```
`tileset.passages.xsize()` is 1168, `tilepicker_data.xsize() * tilepicker_data.ysize()` is 808.

This whole loop seems pretty inefficient anyway, `Table1` and `Table2` both expose `as_slice` methods is it possible you could use [slice.copy_from_slice](https://doc.rust-lang.org/stable/std/primitive.slice.html#method.copy_from_slice)?

white-axe replied:
Looks like this is being caused by tiles and events on some maps having tile IDs larger than the tileset size. I'll have to add some code to the tile rendering as well to deal with this then.

white-axe committed:
Merge branch 'dev' into collision

white-axe committed:
Fix crash when map has tile IDs larger than tileset size

white-axe committed:
Fix crash when tileset `passages` is smaller than the tileset

white-axe committed:
Use `copy_from_slice` to calculate tilepicker passages

melody-rs commented on crates/graphics/src/collision/instance.rs:
```diff
@@ -0,0 +1,204 @@
+
+        render_pass.set_vertex_buffer(1, self.instance_buffer.slice(start..end));
+
+        render_pass.draw(0..12, 0..count);
```
This is only drawing the same 12 vertices, why does `calculate_vertices` create more than 12?

white-axe replied:
I guess I misunderstood somewhere how instancing works.

melody-rs replied:
That's fair! Graphics programming is pretty confusing

white-axe committed:
Remove extra vertices from collision shader

white-axe commented:
Looks like the collision calculation is wrong in certain edge cases. I'll have to investigate.

melody-rs commented on crates/graphics/src/collision/instance.rs:
```diff
@@ -0,0 +1,201 @@
+            .collect_vec()
+    }
+
+    fn calculate_vertices() -> Vec<Vertex> {
```
Could you change this to return `[Vertex; 12]` instead?

white-axe committed:
Collision shader `calculate_vertices` now returns an array instead of a vector

white-axe committed:
Fix collision calculation edge cases

white-axe commented:
Okay, should be fixed now. It wasn't properly handling blank tiles and tiles with the X passage setting.

melody-rs reviewed (approved):
Looks good to me!

white-axe committed:
Simplify `calculate_passage`
